File: cpp/015. 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> ret;
        if(nums.empty())
            return ret;
        sort(nums.begin(), nums.end());
        int target, front, back;
        for(int i = 0; i < nums.size(); i++){
            target = -nums[i];
            front = i + 1;
            back = nums.size() - 1;
            while(front < back){
                int sum = nums[front] + nums[back];
                if(sum > target)
                    back--;
                else if(sum < target)
                    front++;
                else{
                    vector<int> v(3);
                    v[0] = nums[i];
                    v[1] = nums[front];
                    v[2] = nums[back];
                    ret.push_back(v);
                    while(front < back && nums[front] == v[1])
                        front++;
                    while(front < back && nums[back] == v[2])
                        back--;
                }
            }
            while(nums[i + 1] == nums[i])
                i++;
        }
        return ret;
    }
};

File: java/Array/015. 3Sum.java
import java.util.*;

public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        int target, l, r;
        List<List<Integer>> res = new ArrayList<>();

        for(int i = 0; i < nums.length - 2; i++) {
            if(i > 0 && nums[i] == nums[i - 1]) continue;
            target = -1 * nums[i];
            l = i + 1;
            r = nums.length - 1;

            while(l < r) {
                int sum = nums[l] + nums[r];
                if(sum < target) l++;
                else if(sum > target) r--;
                else {
                    List<Integer> triplet = Arrays.asList(nums[i], nums[l], nums[r]);
                    res.add(triplet);

                    l++;
                    r--;
                    // Avoid duplication
                    while(l < r && nums[l] == nums[l - 1]) l++;
                    while(l < r && nums[r] == nums[r + 1]) r--;
                }
            }

        }
        return res;
    }
}
